graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -312,20 +312,40 @@ def _get_recurse_directive_depth(field_name, field_directives):
     return recurse_depth
 
 
-def _validate_recurse_directive_types(current_schema_type, field_schema_type):
-    """Perform type checks on the enclosing type and the recursed type for a recurse directive."""
-    has_union_type = isinstance(field_schema_type, GraphQLUnionType)
-    is_same_type = current_schema_type.is_same_type(field_schema_type)
+def _validate_recurse_directive_types(current_schema_type, field_schema_type, context):
+    """Perform type checks on the enclosing type and the recursed type for a recurse directive.
+
+    Args:
+        current_schema_type: GraphQLType, the schema type at the current location
+        field_schema_type: GraphQLType, the schema type at the inner scope
+        context: dict, various per-compilation data (e.g. declared tags, whether the current block
+                 is optional, etc.). May be mutated in-place in this function!
+    """
+    # Get the set of all allowed types in the current scope.
+    equivalent_types_to_field_scope = [
```

Comment:
I'm not a fan of this list being of type `GraphQL type object or None`, can we avoid adding to it if the hints / inverse hints dict doesn't have the key?